### PR TITLE
[core] Fix the null function issues in ObjectManager

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
Once moved, the original function should be nullptr. This is triggered in windows with std20 flags turned on.
This PR fixed the issue.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/object_manager/object_manager.cc [hunk 1]
```diff
@@ -90,9 +88,7 @@ ObjectManager::ObjectManager(
           [this, delete_object_callback = std::move(delete_object_callback)](
               const ObjectID &object_id) {
             main_service_->post(
-                [this,
-                 object_id,
-                 delete_object_callback = std::move(delete_object_callback)]() {
+                [this, object_id, &delete_object_callback]() {
```
Comment (rkooo567) [inline src/ray/object_manager/object_manager.cc:91]: Hmm isn't this GC'ed when it is posted to main service though?

Comment (iycheng) [reply src/ray/object_manager/object_manager.cc:91]: It'll. But I think we call io_context.stop() first. so the task won't run.

File: src/ray/object_manager/object_manager.cc [hunk 2]
```diff
@@ -78,9 +78,7 @@ ObjectManager::ObjectManager(
           [this, add_object_callback = std::move(add_object_callback)](
               const ObjectInfo &object_info) {
             main_service_->post(
-                [this,
-                 object_info,
-                 add_object_callback = std::move(add_object_callback)]() {
+                [this, object_info, &add_object_callback]() {
```
Comment (rkooo567) [inline src/ray/object_manager/object_manager.cc:81]: Prefer just copy here, but I will leave the final decision up to you. 

```
[this, object_info, add_object_callback]() 
```

**Review by rkooo567 — APPROVED** [review]

**Comment by iycheng** [comment]
Synced offline with @rkooo567 
Basically, the capture also has `this` and `this`'s lifecycle is the same as the ref capture. So it's safe here.
I'll keep it as it is.

**Pull request merged**